mediaapi/routing/preview_url.go

```diff
@@ -0,0 +1,400 @@
+			Origin:        cfg.Matrix.ServerName,
+			FileSizeBytes: bytesWritten,
+			Base64Hash:    hash,
+			ContentType:   types.ContentType(response.Header["Content-Type"][0]),
+		},
+		Logger: util.GetLogger(ctx).WithField("Origin", cfg.Matrix.ServerName),
+	}
+
+	// :-\
+	// Look up the media by the file hash. If we already have the file but under a
+	// different media ID then we won't upload the file again - instead we'll just
+	// add a new metadata entry that refers to the same file.
+	existingMetadata, err := db.GetMediaMetadataByHash(
+		ctx, hash, urlMetadata.MediaMetadata.Origin,
+	)
+	if err != nil {
+		fileutils.RemoveDir(tmpDir, urlMetadata.Logger)
+		urlMetadata.Logger.WithError(err).Error("Error querying the database by hash.")
+		resErr := jsonerror.InternalServerError()
+		return nil, &resErr
+	}
+	if existingMetadata != nil {
+		// The file already exists, delete the uploaded temporary file.
+		defer fileutils.RemoveDir(tmpDir, urlMetadata.Logger)
+		// The file already exists. Make a new media ID up for it.
+		mediaID, merr := urlMetadata.generateMediaID(ctx, db)
+		if merr != nil {
+			urlMetadata.Logger.WithError(merr).Error("Failed to generate media ID for existing file")
+			resErr := jsonerror.InternalServerError()
+			return nil, &resErr
+		}
+
+		// Then amend the upload metadata.
+		urlMetadata.MediaMetadata = &types.MediaMetadata{
+			MediaID:           mediaID,
+			Origin:            urlMetadata.MediaMetadata.Origin,
+			ContentType:       urlMetadata.MediaMetadata.ContentType,
+			FileSizeBytes:     urlMetadata.MediaMetadata.FileSizeBytes,
+			CreationTimestamp: urlMetadata.MediaMetadata.CreationTimestamp,
+			UploadName:        urlMetadata.MediaMetadata.UploadName,
+			Base64Hash:        hash,
+			UserID:            urlMetadata.MediaMetadata.UserID,
+		}
+	} else {
+		// The file doesn't exist. Update the request metadata.
+		urlMetadata.MediaMetadata.FileSizeBytes = bytesWritten
+		urlMetadata.MediaMetadata.Base64Hash = hash
+		urlMetadata.MediaMetadata.MediaID, err = urlMetadata.generateMediaID(ctx, db)
+		if err != nil {
+			fileutils.RemoveDir(tmpDir, urlMetadata.Logger)
+			urlMetadata.Logger.WithError(err).Error("Failed to generate media ID for new download")
+			resErr := jsonerror.InternalServerError()
+			return nil, &resErr
+		}
+	}
+
+	urlMetadata.Logger = urlMetadata.Logger.WithField("media_id", urlMetadata.MediaMetadata.MediaID)
+	urlMetadata.Logger.WithFields(log.Fields{
+		"Base64Hash":    urlMetadata.MediaMetadata.Base64Hash,
+		"UploadName":    urlMetadata.MediaMetadata.UploadName,
+		"FileSizeBytes": urlMetadata.MediaMetadata.FileSizeBytes,
+		"ContentType":   urlMetadata.MediaMetadata.ContentType,
+	}).Info("File downloaded")
+
+	if resErr := urlMetadata.storeFileAndMetadata(ctx, tmpDir, cfg.AbsBasePath, db); resErr != nil {
+		return nil, resErr
+	}
+
+	return urlMetadata, nil
+}
+
+func (m *mediaInfo) generateMediaID(ctx context.Context, db storage.Database) (types.MediaID, error) {
```

Comment:
This function will race with multiple uploads. If 2 people upload a file at the same time it's possible that they will be allocated the same media ID which will then be checked in the database, see it's free and then overwrite each other later on. This is however unlikely given the 32 bytes of entropy used here so maybe that's okay.